lib/ui/painting.dart

```diff
@@ -4935,26 +4591,1632 @@ class PictureRecorder extends NativeFieldWrapperClass2 {
+      } else {
+        newData = ByteData(4 * _newSize(_dataInts.length, _minDataSize, _maxDataGrow, dataNeeded));
+      }
+      final Uint32List newInts = newData.buffer.asUint32List();
+      newInts.setRange(0, _numData, _dataInts);
+      _data = newData;
+      _dataInts = newInts;
+      _dataFloats = newData.buffer.asFloat32List();
+    }
+    assert(_numOps < _ops.length);
+    _ops[_numOps++] = op.index;
+  }
+
+  void _addInt(int value) {
+    assert(_numData < _dataInts.length);
+    _dataInts[_numData++] = value;
+  }
+
+  void _addScalar(double value) {
+    assert(_numData < _dataFloats.length);
+    _dataFloats[_numData++] = value;
+  }
+
+  void _addScalar2(double v1, double v2) {
+    assert(_numData + 2 <= _dataFloats.length);
+    _dataFloats[_numData++] = v1;
+    _dataFloats[_numData++] = v2;
+  }
+
+  void _addScalar3(double v1, double v2, double v3) {
+    assert(_numData + 3 <= _dataFloats.length);
+    _dataFloats[_numData++] = v1;
+    _dataFloats[_numData++] = v2;
+    _dataFloats[_numData++] = v3;
+  }
+
+  static const int _nOffsetData = 2;
+  void _addOffset(Offset offset) {
+    assert(_numData + _nOffsetData <= _dataFloats.length);
+    _dataFloats[_numData++] = offset.dx;
+    _dataFloats[_numData++] = offset.dy;
+  }
+
+  static const int _nRectData = 4;
+  void _addRect(Rect r) {
+    assert(_numData + _nRectData <= _dataFloats.length);
+    _dataFloats[_numData++] = r.left;
+    _dataFloats[_numData++] = r.top;
+    _dataFloats[_numData++] = r.right;
+    _dataFloats[_numData++] = r.bottom;
+  }
+
+  static const int _nRoundRectData = 12;
+  void _addRRect(RRect rrect) {
+    assert(_numData + _nRoundRectData <= _dataFloats.length);
+    _dataFloats[_numData++] = rrect.left;
+    _dataFloats[_numData++] = rrect.top;
+    _dataFloats[_numData++] = rrect.right;
+    _dataFloats[_numData++] = rrect.bottom;
+    // SkRRect Radii order is UL, UR, LR, LL as per SkRRect::Corner indices
+    _dataFloats[_numData++] = rrect.tlRadiusX;
+    _dataFloats[_numData++] = rrect.tlRadiusY;
+    _dataFloats[_numData++] = rrect.trRadiusX;
+    _dataFloats[_numData++] = rrect.trRadiusY;
+    _dataFloats[_numData++] = rrect.brRadiusX;
+    _dataFloats[_numData++] = rrect.brRadiusY;
+    _dataFloats[_numData++] = rrect.blRadiusX;
+    _dataFloats[_numData++] = rrect.blRadiusY;
+  }
+
+  void _addInt32List(Int32List data) {
+    final int len = data.length;
+    assert(_numData + len + 1 <= _dataInts.length);
+    _dataInts[_numData++] = len;
+    _dataInts.setRange(_numData, _numData + len, data);
+    _numData += len;
+  }
+
+  void _addFloat32List(Float32List data) {
+    final int len = data.length;
+    assert(_numData + len + 1 <= _dataFloats.length);
+    _dataInts[_numData++] = len;
+    _dataFloats.setRange(_numData, _numData + len, data);
+    _numData += len;
+  }
+
+  void _addImageData(Image image) {
+    _objData.add(image._image);
+  }
+
+  void _addPathData(Path path) {
+    _objData.add(path);
```

Comment:
Not necessarily for this PR, but for all of these I'm curious about where and how we'd cache/reuse objects that are identical.

For this PR: if we can change the list to be `List<NativeFieldWrapperClass2>`, we can probably do away with all of these methods and just instead directly add to the list - WDYT?